### PR TITLE
Multi-model and multi-gpu support

### PR DESCRIPTION
**Pull request by alexandonian**
This branch introduces several new features that facilitate multi-model and multi-gpu training/validation under a single, shared dataprovider. This branch also adds support for custom training/validation routines and increased control over pre-training checks (e.g. git repo status). Note: the outward facing API/interface actually remains unchanged.

### CODE REVIEW

File: tfutils/base.py [hunk 1]
```diff
@@ -68,8 +86,20 @@
...
-DEFAULT_LOAD_PARAMS = frozendict({'do_restore': True, 'load_param_dict': None, 'load_step': True, 'from_ckpt': None})
+DEFAULT_PARAMS = frozendict({
+    'dont_run': False,
+    'skip_check': False,
+    'model_params': {},
+    'train_params': {},
+    'validation_params': {},
+    'log_device_placement': False,
+    'inter_op_parallelism_threads': 40,
+    'save_params': dict(DEFAULT_SAVE_PARAMS),
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian should we be using frozendict here? and in similar places in the defaults?

File: tfutils/base.py [hunk 2]
```diff
@@ -216,17 +242,34 @@ def __init__(self,
         self.conn = pymongo.MongoClient(host=self.host, port=self.port)
         self.conn.server_info()
         self.collfs = gridfs.GridFS(self.conn[self.dbname], self.collname)
-        recent_name = '_'.join([self.dbname, self.collname, self.exp_id, '__RECENT'])
+
+        if 'port_rec' in save_params:
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian what is the "port_rec" variable for? it's not documented?

File: tfutils/base.py [hunk 3]
```diff
@@ -484,12 +497,18 @@ def save(self, train_res=None, valid_res=None, step=None, validation_only=False)
         rec['step'] = step
 
         if len(train_res) > 0:
-            # TODO: also include error rate of the train set to monitor overfitting
+            # TODO: also include error rate of the train set to monitor
+            # overfitting
             message = 'Step {} ({:.0f} ms) -- '.format(step, 1000 * duration)
-            msg2 = ['{}: {:.4f}'.format(k, v) for k, v in train_res.items() if k != 'optimizer' and k not in self.save_to_gfs]
+            msg2 = ['{}: {:.4f}'.format(k, v) for k, v in train_res.items()
+                    if k not in ['optimizer', 'grads', 'zero_grad'] and k not in self.save_to_gfs]
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian actually maybe we should print out a summary of the gradients.   

btw, what is the "zero_grad" object?

File: tfutils/base.py [hunk 4]
```diff
@@ -198,15 +222,17 @@ def __init__(self,
...
+        if 'query' in load_params and not load_params['query'] is None and 'exp_id' in load_params['query']:
+            self.sameloc = self.sameloc & (
+                load_params['query']['exp_id'] == self.exp_id)
 
         for _k in ['do_save', 'save_metrics_freq', 'save_valid_freq', 'cache_filters_freq',
                    'save_filters_freq', 'save_initial_filters', 'save_to_gfs']:
             setattr(self, _k, save_params.get(_k, DEFAULT_SAVE_PARAMS[_k]))
 
-        for _k in ['do_restore', 'load_param_dict', 'load_step', 'from_ckpt']:
+        for _k in ['do_restore']:
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian why did you delete these variables from being set by default?

File: tfutils/base.py [hunk 5]
```diff
@@ -274,76 +318,47 @@ def load_rec(self):
...
-                raise Exception('You specified load parameters but no record was found with the given spec.')
+                raise Exception(
+                    'You specified load parameters but no record was found with the given spec.')
         self.load_data = load
 
-    def initialize(self):
-        """
-        Fetches record then uses tf's saver.restore
-        """
+    def initialize(self, no_scratch=False):
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian what is the no_scratch variable?

File: tfutils/base.py [hunk 6]
```diff
@@ -718,16 +764,35 @@ def test(sess,
...
-    return valid_results_summary, dbinterface.outrecs
+    # Collect args in a dict of lists
+    test_args = {
+        'sess': sess,
+        'queues': queues,
+        'dbinterface': dbinterface,
+        'validation_targets': validation_targets,
+        'save_intermediate_freq': save_intermediate_freq}
+
+    _ttargs = [{key: value[i] for (key, value) in test_args.items()}
```
Comment (yamins81) [inline tfutils/base.py]: @alexandonian what is sess a list?

Comment (yamins81) [inline tfutils/base.py]: is it possible to have different sessions here?

File: tfutils/base.py [hunk 7]
```diff
@@ -736,78 +801,147 @@ def test_from_params(load_params,
...
+
+    A training loop that performs minibatching with `num_minibatches`
+    minibatches.
+
+    Args:
+        sess (tf.Session): Current tensorflow session.
+        train_targets (dict): Target operations to be evaluated by `sess.run`.
+            By default, `base.train_from_params` inserts the following
+            targets to facilitate minibatching:
+            - `grads` (tf.Variable): Stores accumulated gradients.
```
Comment (yamins81) [inline tfutils/base.py]: should (a) the names of these special ops be more exotic to avoid stepping on them? like "grads" or something? and (b) should this function check that the relevant ops exist?

File: tfutils/base.py [hunk 8]
```diff
@@ -816,49 +950,112 @@ def train(sess,
...
+
+    # Run training
+    while any(step < num_step for (step, num_step) in zip(steps, num_steps)):
+
+        # Make a single call to sess.run to produce a list of results.
+        # TODO: call default train_loop which has minibatch option and is
+        # defined in base.py
+
+        start_time_step = time.time()
+        if train_loop is not None:
```
Comment (yamins81) [inline tfutils/base.py]: there shouldn't be a case here.   there should be a default train_loop function, which is what is called unless the train_loop is specified.   moreover,  arguments to the function be allowed to be specified by the user

Comment (yamins81) [inline tfutils/base.py]: this is especially true since train_loop with never be None.  the 'else' case is never run.

File: tfutils/base.py [hunk 9]
```diff
@@ -816,49 +950,112 @@ def train(sess,
...
+    for trarg in trargs:
+        stop_queues(trarg['sess'], trarg['queues'],
+                    trarg['coord'], trarg['threads'])
+        trarg['dbinterface'].sync_with_host()
+        res.append(trarg['dbinterface'].outrecs)
 
-    stop_queues(sess, queues, coord, threads)
-    dbinterface.sync_with_host()
-    return dbinterface.outrecs
+    trargs[0]['sess'].close()
```
Comment (yamins81) [inline tfutils/base.py]: how come we just get away with closing trargs[0]['sess']?  are we assuming the sess is the same for all trargs elements

File: tfutils/tests/test.py [hunk 10]
```diff
@@ -12,46 +13,135 @@
...
     conn = pm.MongoClient(host=testhost,
                           port=testport)
-    conn.drop_database(testdbname)
-    nm = testdbname + '_' + testcol + '_training0'
-    [conn.drop_database(x) for x in conn.database_names() if x.startswith(nm) and '___RECENT' in x]
-    nm = testdbname + '_' + testcol + '_training1'
-    [conn.drop_database(x) for x in conn.database_names() if x.startswith(nm) and '___RECENT' in x]
+
+    # delete old collection if it exists
+    # conn[testdbname][testcol + '.files'].drop()
```
Comment (yamins81) [inline tfutils/tests/test.py]: so how are these tests supposed to be ran? in what order?   if this deletion code is not run, the tests won't run a second time correctly.  where is the deletion supposed to happen?   is there a test script?

File: tfutils/base.py (hunk 1 above)
Comment (alexandonian) [inline tfutils/base.py]: will do.

File: tfutils/base.py (hunk 4 above)
Comment (yamins81) [inline tfutils/base.py]: We need to:

1.  do a careful diff rel. master and make sure nothing else like this has escaped our attention

2. the initialize fn has to really be cleaned up.   There are three needs:

    - from_ckpt, which will load from a file as opposed to from the DB.  This should be done with the same API as @anayebi wrote but should be cleaned up to ensure that it is orthogonal from the next few issues. 
    - var_list, which will specify which things from the loaded checkpoint to actually use as opposed to initialize newly.   this should probably be a list of strings or a regex or callable which returns strings? (but think about it)
    - mapping_dict, which says, for various names of what are loaded and used, what their new names are

All three of these issues should be totally orthogonal, so that eg. specifying the mapping shouldn't get in the way of loading other variables that are not renamed.

Comment (yamins81) [inline tfutils/base.py]: this has to be checked out with @ewang314 @anayebi @chengxuz

File: tfutils/base.py (hunk 2 above)
Comment (yamins81) [inline tfutils/base.py]: @chengxuz what does "port_rec" do?   @alexandonian can we clean this up to make the code more clear?

File: tfutils/base.py (hunk 5 above)
Comment (yamins81) [inline tfutils/base.py]: @chengxuz can you and alex chat about *why* this is being done and make sure it's a clean and simple use case?  and not covered / interacting with other things?

File: tfutils/base.py (hunk 3 above)
Comment (yamins81) [inline tfutils/base.py]: is there some less "ad hoc" way of specifying this so that (e.g.) other code like this save method doesn't have to know about the details of how the optimizer is up?

File: tfutils/base.py (hunk 6 above)
Comment (yamins81) [inline tfutils/base.py]: because sess is always the same across the list, just break it out and pass one session everywhere

File: tfutils/base.py (hunk 7 above)
Comment (yamins81) [inline tfutils/base.py]: specifically, amke sure that the optimizer and `___grads___` present and issue error otherwise

File: tfutils/base.py [hunk 11]
```diff
@@ -816,49 +950,112 @@ def train(sess,
...
+            Callable that specifies a custom training loop
         - train_targets (dict of tensorflow nodes)
             Targets to train. One item in this dict must be "optimizer" or similar
             to make anything happen
+        - num_minibatches (int)
+            How many minibatches to use to before applying gradient update.
         - num_steps (int)
             How many steps to train to before quitting
-        - valid_targets (dict of tensorflow objects, default: None)
+        - validdation_targets (dict of tensorflow objects, default: None)
```
Comment (yamins81) [inline tfutils/base.py]: spelling error

File: tfutils/tests/test.py (hunk 10 above)
Comment (yamins81) [inline tfutils/tests/test.py]: actually yeah let's do a real test suite using nose or nose2 or whatever, and with proper setup/teardown

File: tfutils/tests/test.py [hunk 12]
```diff
@@ -12,46 +13,135 @@
...
+    # nm = testdbname + '_' + testcol + '_training0'
+    # [conn.drop_database(x) for x in conn.database_names() if x.startswith(nm) and '___RECENT' in x]
+    # nm = testdbname + '_' + testcol + '_training1'
+    # [conn.drop_database(x) for x in conn.database_names() if x.startswith(nm) and '___RECENT' in x]
```
Comment (yamins81) [inline tfutils/tests/test.py]: let's delete these commented out lines

File: tutorials/.ipynb_checkpoints/Notebook Corresponding to Training Test-checkpoint.ipynb [hunk 13]
```diff
@@ -0,0 +1,294 @@
+{
```
Comment (yamins81) [inline tutorials/.ipynb_checkpoints/Notebook Corresponding to Training Test-checkpoint.ipynb]: let's delete the directory called ".ipynb_checkpoints" as the only notebook in this is already in the enclosing folder

File: tfutils/base.py (hunk 2 above)
Comment (chengxuz) [inline tfutils/base.py]: Feel free to delete every thing related to "port_rec", it's used for sending another copy of records to another database, trying to overcome the problem that the database was too slow at that time

File: tfutils/base.py (hunk 1 above)
Comment (alexandonian) [inline tfutils/base.py]: Fixed.

File: tfutils/base.py (hunk 2 above)
Comment (alexandonian) [inline tfutils/base.py]: Fixed.

File: tfutils/base.py (hunk 11 above)
Comment (alexandonian) [inline tfutils/base.py]: Fixed.

File: tutorials/.ipynb_checkpoints/Notebook Corresponding to Training Test-checkpoint.ipynb (hunk 13 above)
Comment (alexandonian) [inline tutorials/.ipynb_checkpoints/Notebook Corresponding to Training Test-checkpoint.ipynb]: Fixed.

File: tfutils/base.py (hunk 6 above)
Comment (alexandonian) [inline tfutils/base.py]: Fixed.

File: tfutils/tests/test.py (hunk 12 above)
Comment (alexandonian) [inline tfutils/tests/test.py]: Fixed, for now.
